In [1]:
# Required Python Machine learning Packages
import pandas as pd
import numpy as np

In [2]:
bank_df = pd.read_csv('bank.csv',header = 0, delimiter=";") 
#read csv file and use delimiter for read csv in that every value seperated by ";"
bank_df

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

In [197]:
print(bank_df.dtypes)
print(bank_df.shape)
print(bank_df.describe())

age                 int64
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome           object
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                  object
dtype: object
(41188, 21)
               age      duration      campaign         pdays      previous  \
count  41188.00000  41188.000000  41188.000000  41188.000000  41188.000000   
mean      40.02406    258.285010      2.567593    962.475454      0.172963   
std       10.42125    259.279249      2.770014    186.910907      0.494901   
min       17.00000      0.000000      1.000000      0.000000      0.000000   
25%       32.00000    102.000000      

In [198]:
# for see the null value
bank_df.isnull().sum()

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp.var.rate      0
cons.price.idx    0
cons.conf.idx     0
euribor3m         0
nr.employed       0
y                 0
dtype: int64

In [199]:
bank_df=bank_df.drop(["contact","month","day_of_week"],axis=1)
bank_df.shape


(41188, 18)

In [200]:
bank_df

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

       duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0           261         1    999         0  nonexistent           1.1   
1           149         1    999         0  nonexistent           1.1   
2           226         1    999         0  nonexistent           1.1   
3           151         1    999         0  nonexistent           1.1   
4           307         1    999         0  nonexistent           1.1   
...         ...       ...    ...       ...          ...           ...   
41183       334         1    999         0  nonexistent          -1.1   
41184       383         1    999         0  nonexistent          -1.1   
41185       189         2    999         0  nonexistent          -1.1   
41186       442         1    999         0  nonexistent          -1.1   
41187       239         3    999         1      failure          -1.1   

       cons.price.idx  cons.conf.idx  euribor3m  nr.employed    y  
0              93.994          -36.4      4.857       5191.0   no  
1              93.994          -36.4      4.857       5191.0   no  
2              93.994          -36.4      4.857       5191.0   no  
3              93.994          -36.4      4.857       5191.0   no  
4              93.994          -36.4      4.857       5191.0   no  
...               ...            ...        ...          ...  ...  
41183          94.767          -50.8      1.028       4963.6  yes  
41184          94.767          -50.8      1.028       4963.6   no  
41185          94.767          -50.8      1.028       4963.6   no  
41186          94.767          -50.8      1.028       4963.6  yes  
41187          94.767          -50.8      1.028       4963.6   no  

[41188 rows x 18 columns]

# Missing Value Analysis

In [201]:
#some variable contain "unknown" value so we should change it.
bank_df=bank_df.replace('unknown',np.nan)

In [202]:
bank_df

age          job  marital            education default housing loan  \
0       56    housemaid  married             basic.4y      no      no   no   
1       57     services  married          high.school     NaN      no   no   
2       37     services  married          high.school      no     yes   no   
3       40       admin.  married             basic.6y      no      no   no   
4       56     services  married          high.school      no      no  yes   
...    ...          ...      ...                  ...     ...     ...  ...   
41183   73      retired  married  professional.course      no     yes   no   
41184   46  blue-collar  married  professional.course      no      no   no   
41185   56      retired  married    university.degree      no     yes   no   
41186   44   technician  married  professional.course      no      no   no   
41187   74      retired  married  professional.course      no     yes   no   

       duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0           261         1    999         0  nonexistent           1.1   
1           149         1    999         0  nonexistent           1.1   
2           226         1    999         0  nonexistent           1.1   
3           151         1    999         0  nonexistent           1.1   
4           307         1    999         0  nonexistent           1.1   
...         ...       ...    ...       ...          ...           ...   
41183       334         1    999         0  nonexistent          -1.1   
41184       383         1    999         0  nonexistent          -1.1   
41185       189         2    999         0  nonexistent          -1.1   
41186       442         1    999         0  nonexistent          -1.1   
41187       239         3    999         1      failure          -1.1   

       cons.price.idx  cons.conf.idx  euribor3m  nr.employed    y  
0              93.994          -36.4      4.857       5191.0   no  
1              93.994          -36.4      4.857       5191.0   no  
2              93.994          -36.4      4.857       5191.0   no  
3              93.994          -36.4      4.857       5191.0   no  
4              93.994          -36.4      4.857       5191.0   no  
...               ...            ...        ...          ...  ...  
41183          94.767          -50.8      1.028       4963.6  yes  
41184          94.767          -50.8      1.028       4963.6   no  
41185          94.767          -50.8      1.028       4963.6   no  
41186          94.767          -50.8      1.028       4963.6  yes  
41187          94.767          -50.8      1.028       4963.6   no  

[41188 rows x 18 columns]

In [203]:
# in place of object datatype "unknown" value is replace with mode of same column.
# in place of int and float datatype "unknown" value is replace with mean of same column.

for x in bank_df.columns:
    if bank_df[x].dtype=='object' or bank_df[x].dtype=='bool':
        bank_df[x].fillna(bank_df[x].mode()[0],inplace=True) # some time mode is bimodal there for we want to select 1st mode use mode[0]
    elif bank_df[x].dtype=='int64' or bank_df[x].dtype=='float64':
        bank_df[x].fillna(round(bank_df[x].mean()),inplace=True)

In [204]:
bank_df

age          job  marital            education default housing loan  \
0       56    housemaid  married             basic.4y      no      no   no   
1       57     services  married          high.school      no      no   no   
2       37     services  married          high.school      no     yes   no   
3       40       admin.  married             basic.6y      no      no   no   
4       56     services  married          high.school      no      no  yes   
...    ...          ...      ...                  ...     ...     ...  ...   
41183   73      retired  married  professional.course      no     yes   no   
41184   46  blue-collar  married  professional.course      no      no   no   
41185   56      retired  married    university.degree      no     yes   no   
41186   44   technician  married  professional.course      no      no   no   
41187   74      retired  married  professional.course      no     yes   no   

       duration  campaign  pdays  previous     poutcome  emp.var.rate  \
0           261         1    999         0  nonexistent           1.1   
1           149         1    999         0  nonexistent           1.1   
2           226         1    999         0  nonexistent           1.1   
3           151         1    999         0  nonexistent           1.1   
4           307         1    999         0  nonexistent           1.1   
...         ...       ...    ...       ...          ...           ...   
41183       334         1    999         0  nonexistent          -1.1   
41184       383         1    999         0  nonexistent          -1.1   
41185       189         2    999         0  nonexistent          -1.1   
41186       442         1    999         0  nonexistent          -1.1   
41187       239         3    999         1      failure          -1.1   

       cons.price.idx  cons.conf.idx  euribor3m  nr.employed    y  
0              93.994          -36.4      4.857       5191.0   no  
1              93.994          -36.4      4.857       5191.0   no  
2              93.994          -36.4      4.857       5191.0   no  
3              93.994          -36.4      4.857       5191.0   no  
4              93.994          -36.4      4.857       5191.0   no  
...               ...            ...        ...          ...  ...  
41183          94.767          -50.8      1.028       4963.6  yes  
41184          94.767          -50.8      1.028       4963.6   no  
41185          94.767          -50.8      1.028       4963.6   no  
41186          94.767          -50.8      1.028       4963.6  yes  
41187          94.767          -50.8      1.028       4963.6   no  

[41188 rows x 18 columns]

# Label Encoding wherever required

In [205]:
colname=[]
for i in bank_df.columns:
    if bank_df[i].dtype=='object':
        colname.append(i)
colname

['job', 'marital', 'education', 'default', 'housing', 'loan', 'poutcome', 'y']

In [206]:
# For preprocessing the data
from sklearn.preprocessing import LabelEncoder

le=LabelEncoder()

for i in colname:
    bank_df[i]=le.fit_transform(bank_df[i])

In [207]:
bank_df

age  job  marital  education  default  housing  loan  duration  \
0       56    3        1          0        0        0     0       261   
1       57    7        1          3        0        0     0       149   
2       37    7        1          3        0        1     0       226   
3       40    0        1          1        0        0     0       151   
4       56    7        1          3        0        0     1       307   
...    ...  ...      ...        ...      ...      ...   ...       ...   
41183   73    5        1          5        0        1     0       334   
41184   46    1        1          5        0        0     0       383   
41185   56    5        1          6        0        1     0       189   
41186   44    9        1          5        0        0     0       442   
41187   74    5        1          5        0        1     0       239   

       campaign  pdays  previous  poutcome  emp.var.rate  cons.price.idx  \
0             1    999         0         1           1.1          93.994   
1             1    999         0         1           1.1          93.994   
2             1    999         0         1           1.1          93.994   
3             1    999         0         1           1.1          93.994   
4             1    999         0         1           1.1          93.994   
...         ...    ...       ...       ...           ...             ...   
41183         1    999         0         1          -1.1          94.767   
41184         1    999         0         1          -1.1          94.767   
41185         2    999         0         1          -1.1          94.767   
41186         1    999         0         1          -1.1          94.767   
41187         3    999         1         0          -1.1          94.767   

       cons.conf.idx  euribor3m  nr.employed  y  
0              -36.4      4.857       5191.0  0  
1              -36.4      4.857       5191.0  0  
2              -36.4      4.857       5191.0  0  
3              -36.4      4.857       5191.0  0  
4              -36.4      4.857       5191.0  0  
...              ...        ...          ... ..  
41183          -50.8      1.028       4963.6  1  
41184          -50.8      1.028       4963.6  0  
41185          -50.8      1.028       4963.6  0  
41186          -50.8      1.028       4963.6  1  
41187          -50.8      1.028       4963.6  0  

[41188 rows x 18 columns]

In [208]:
#to see, in Y variable, which category have which number.
le_name_mapping = dict(zip(le.classes_, le.transform(le.classes_)))
print('Feature', i)
print('mapping', le_name_mapping)

Feature y
mapping {'no': 0, 'yes': 1}


In [209]:
print(bank_df.dtypes)

age                 int64
job                 int32
marital             int32
education           int32
default             int32
housing             int32
loan                int32
duration            int64
campaign            int64
pdays               int64
previous            int64
poutcome            int32
emp.var.rate      float64
cons.price.idx    float64
cons.conf.idx     float64
euribor3m         float64
nr.employed       float64
y                   int32
dtype: object


In [210]:
X = bank_df.values[:,0:-1] #select all columns only last column is not consider,
Y = bank_df.values[:,-1] # select only last column,

In [211]:
X

array([[ 5.6000e+01,  3.0000e+00,  1.0000e+00, ..., -3.6400e+01,
         4.8570e+00,  5.1910e+03],
       [ 5.7000e+01,  7.0000e+00,  1.0000e+00, ..., -3.6400e+01,
         4.8570e+00,  5.1910e+03],
       [ 3.7000e+01,  7.0000e+00,  1.0000e+00, ..., -3.6400e+01,
         4.8570e+00,  5.1910e+03],
       ...,
       [ 5.6000e+01,  5.0000e+00,  1.0000e+00, ..., -5.0800e+01,
         1.0280e+00,  4.9636e+03],
       [ 4.4000e+01,  9.0000e+00,  1.0000e+00, ..., -5.0800e+01,
         1.0280e+00,  4.9636e+03],
       [ 7.4000e+01,  5.0000e+00,  1.0000e+00, ..., -5.0800e+01,
         1.0280e+00,  4.9636e+03]])

In [212]:
Y

array([0., 0., 0., ..., 0., 1., 0.])

In [213]:
X.shape

(41188, 17)

In [214]:
Y.shape

(41188,)

In [215]:
Y=Y.astype(int)

In [216]:
# Standardize the data

In [217]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)
print(X) 

[[ 1.53303429 -0.17929992 -0.27980323 ...  0.88644656  0.71245988
   0.33167991]
 [ 1.62899323  0.94757962 -0.27980323 ...  0.88644656  0.71245988
   0.33167991]
 [-0.29018564  0.94757962 -0.27980323 ...  0.88644656  0.71245988
   0.33167991]
 ...
 [ 1.53303429  0.38413985 -0.27980323 ... -2.22495344 -1.49518647
  -2.8156966 ]
 [ 0.38152696  1.5110194  -0.27980323 ... -2.22495344 -1.49518647
  -2.8156966 ]
 [ 3.26029527  0.38413985 -0.27980323 ... -2.22495344 -1.49518647
  -2.8156966 ]]


In [218]:
#split the data in train and test by 80% and 20%

from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2,
random_state=10)

In [219]:
X_test.shape

(8238, 17)

In [220]:
Y_test.shape

(8238,)

In [221]:
X_train.shape

(32950, 17)

In [222]:
Y_train.shape

(32950,)

# Selecting important features based on Random Forest

In [223]:
#model building

from sklearn.ensemble import RandomForestClassifier

model_RandomForest=RandomForestClassifier(n_estimators=50, random_state=10)

#fit the model on the data and predict the values
model_RandomForest.fit(X_train,Y_train)

Y_pred=model_RandomForest.predict(X_test)

In [224]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
print(classification_report(Y_test,Y_pred))

[[7057  253]
 [ 473  455]]
0.9118718135469774
              precision    recall  f1-score   support

           0       0.94      0.97      0.95      7310
           1       0.64      0.49      0.56       928

    accuracy                           0.91      8238
   macro avg       0.79      0.73      0.75      8238
weighted avg       0.90      0.91      0.91      8238



In [225]:
print(list(zip(bank_df.columns[0:-1],model_RandomForest.feature_importances_)))

[('age', 0.10351796015362934), ('job', 0.05041948113906849), ('marital', 0.023910101383509897), ('education', 0.04217621707843848), ('default', 9.725374461268737e-10), ('housing', 0.019698475257294967), ('loan', 0.01350650479869965), ('duration', 0.3371795126067871), ('campaign', 0.04623312334378058), ('pdays', 0.03358133774973585), ('previous', 0.013831438777205886), ('poutcome', 0.02216085269241215), ('emp.var.rate', 0.020649914673076007), ('cons.price.idx', 0.025707751656677092), ('cons.conf.idx', 0.027547441089038), ('euribor3m', 0.1554922766072508), ('nr.employed', 0.06438761002085822)]


In [226]:
#feature importances value of default variable is approx 0% there for drop it

bank_df=bank_df.drop(["default"],axis=1)
bank_df.shape

(41188, 17)

In [227]:
#assign the X variable
X = bank_df.values[:,0:-1]
Y = bank_df.values[:,-1]
Y=Y.astype(int)

# Standardize the data using the anyone of the scalers 

In [228]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit(X)
X = scaler.transform(X)
print(X) 

[[ 1.53303429 -0.17929992 -0.27980323 ...  0.88644656  0.71245988
   0.33167991]
 [ 1.62899323  0.94757962 -0.27980323 ...  0.88644656  0.71245988
   0.33167991]
 [-0.29018564  0.94757962 -0.27980323 ...  0.88644656  0.71245988
   0.33167991]
 ...
 [ 1.53303429  0.38413985 -0.27980323 ... -2.22495344 -1.49518647
  -2.8156966 ]
 [ 0.38152696  1.5110194  -0.27980323 ... -2.22495344 -1.49518647
  -2.8156966 ]
 [ 3.26029527  0.38413985 -0.27980323 ... -2.22495344 -1.49518647
  -2.8156966 ]]


In [229]:
#split data as train and test
from sklearn.model_selection import train_test_split

X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2,
random_state=10)

In [230]:
X_test.shape

(8238, 16)

In [231]:
Y_train.shape

(32950,)

In [232]:
X_train.shape

(32950, 16)

In [233]:
Y_test.shape

(8238,)

In [234]:
#after tunning recheck accuracy

model_RandomForest=RandomForestClassifier(n_estimators=50, random_state=10)

#fit the model on the data and predict the values
model_RandomForest.fit(X_train,Y_train)

Y_pred=model_RandomForest.predict(X_test)


from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
print(classification_report(Y_test,Y_pred))

[[7046  264]
 [ 461  467]]
0.9119932022335518
              precision    recall  f1-score   support

           0       0.94      0.96      0.95      7310
           1       0.64      0.50      0.56       928

    accuracy                           0.91      8238
   macro avg       0.79      0.73      0.76      8238
weighted avg       0.90      0.91      0.91      8238



#  Logistic Regression

In [235]:
#model building for logistic regression.
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression()
classifier.fit(X_train, Y_train)
Y_pred=classifier.predict(X_test)
print(Y_pred)

[0 0 0 ... 0 0 0]


In [236]:
print(list(zip(Y_test, Y_pred)))

[(0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 1), (0, 0), (0, 0), (0, 1), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0),

In [237]:
# generate confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[7111  199]
 [ 585  343]]
Classification report: 
              precision    recall  f1-score   support

           0       0.92      0.97      0.95      7310
           1       0.63      0.37      0.47       928

    accuracy                           0.90      8238
   macro avg       0.78      0.67      0.71      8238
weighted avg       0.89      0.90      0.89      8238

Accuracy of the model:  0.9048312697256615


In [238]:
# for the tunning

In [239]:
Y_pred_prob = classifier.predict_proba(X_test)
print(Y_pred_prob)

[[0.84932631 0.15067369]
 [0.93891455 0.06108545]
 [0.52880827 0.47119173]
 ...
 [0.98446063 0.01553937]
 [0.98706438 0.01293562]
 [0.99231086 0.00768914]]


In [240]:
Y_pred_class=[]
for value in Y_pred_prob[:,1]:
    if value > 0.42: #after tunning select proper thresold values.
        Y_pred_class.append(1)
    else:
        Y_pred_class.append(0)
print(Y_pred_class)

[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [241]:
# GEnerate confusion matrix of logistic regression

from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred_class)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred_class))

acc=accuracy_score(Y_test, Y_pred_class)
print("Accuracy of the model: ",acc)

[[7057  253]
 [ 512  416]]
Classification report: 
              precision    recall  f1-score   support

           0       0.93      0.97      0.95      7310
           1       0.62      0.45      0.52       928

    accuracy                           0.91      8238
   macro avg       0.78      0.71      0.73      8238
weighted avg       0.90      0.91      0.90      8238

Accuracy of the model:  0.9071376547705754


In [242]:

#find the error for different thresold

for a in np.arange(0.3,0.61,0.01):
    predict_mine = np.where(Y_pred_prob[:,1] > a, 1, 0)
    cfm=confusion_matrix(Y_test, predict_mine)
    total_err=cfm[0,1]+cfm[1,0]
    print("Errors at threshold ", a, ":",total_err, " , type 2 error :",
          cfm[1,0]," , type 1 error:", cfm[0,1])

Errors at threshold  0.3 : 778  , type 2 error : 390  , type 1 error: 388
Errors at threshold  0.31 : 779  , type 2 error : 403  , type 1 error: 376
Errors at threshold  0.32 : 777  , type 2 error : 415  , type 1 error: 362
Errors at threshold  0.33 : 772  , type 2 error : 428  , type 1 error: 344
Errors at threshold  0.34 : 765  , type 2 error : 438  , type 1 error: 327
Errors at threshold  0.35000000000000003 : 767  , type 2 error : 453  , type 1 error: 314
Errors at threshold  0.36000000000000004 : 766  , type 2 error : 465  , type 1 error: 301
Errors at threshold  0.37000000000000005 : 769  , type 2 error : 479  , type 1 error: 290
Errors at threshold  0.38000000000000006 : 766  , type 2 error : 483  , type 1 error: 283
Errors at threshold  0.39000000000000007 : 772  , type 2 error : 494  , type 1 error: 278
Errors at threshold  0.4000000000000001 : 771  , type 2 error : 500  , type 1 error: 271
Errors at threshold  0.4100000000000001 : 767  , type 2 error : 508  , type 1 error: 25

 # Adaboost

In [243]:
#predicting using the AdaBoost_Classifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier

model_AdaBoost=AdaBoostClassifier(base_estimator=DecisionTreeClassifier(random_state=10),
n_estimators=50,
random_state=10)
#fit the model on the data and predict the values
model_AdaBoost.fit(X_train,Y_train)
Y_pred=model_AdaBoost.predict(X_test)

In [244]:
from sklearn.metrics import confusion_matrix, accuracy_score,classification_report
#confusion matrix
print(confusion_matrix(Y_test,Y_pred))
print(accuracy_score(Y_test,Y_pred))
print(classification_report(Y_test,Y_pred))

[[6834  476]
 [ 441  487]]
0.8886865744112649
              precision    recall  f1-score   support

           0       0.94      0.93      0.94      7310
           1       0.51      0.52      0.52       928

    accuracy                           0.89      8238
   macro avg       0.72      0.73      0.73      8238
weighted avg       0.89      0.89      0.89      8238



# KNN

In [245]:
#predicting using the KNeighbors_Classifier
from sklearn.neighbors import KNeighborsClassifier
model_KNN=KNeighborsClassifier(n_neighbors=int(np.sqrt(len(X_train))), 
metric='minkowski')

model_KNN.fit(X_train,Y_train)

y_pred=model_KNN.predict(X_test)
print(list(zip(Y_test,Y_pred)))

[(0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 1), (0, 0), (0, 0), (0, 0), (1, 0), (0, 0), (0, 0), (1, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 1), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (0, 0), (1, 1), (0, 0),

In [246]:
#confusion matrix
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[6834  476]
 [ 441  487]]
Classification report: 
              precision    recall  f1-score   support

           0       0.94      0.93      0.94      7310
           1       0.51      0.52      0.52       928

    accuracy                           0.89      8238
   macro avg       0.72      0.73      0.73      8238
weighted avg       0.89      0.89      0.89      8238

Accuracy of the model:  0.8886865744112649


# SVM

In [247]:
#predicting using the svm
from sklearn.svm import SVC
svc_model=SVC(kernel='rbf',C=1,gamma=0.1)
svc_model.fit(X_train, Y_train)
Y_pred=svc_model.predict(X_test)
print(list(Y_pred))

[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 

In [248]:
#generate confusion matrix for svm
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test, Y_pred)
print("Accuracy of the model: ",acc)

[[7105  205]
 [ 594  334]]
Classification report: 
              precision    recall  f1-score   support

           0       0.92      0.97      0.95      7310
           1       0.62      0.36      0.46       928

    accuracy                           0.90      8238
   macro avg       0.77      0.67      0.70      8238
weighted avg       0.89      0.90      0.89      8238

Accuracy of the model:  0.9030104394270454


After tunning we get accuracy score for different alog. is

1. Random forest = 91.19%
2. logistic regression = 90.71%
3. adaboost = 88.86%
4. KNN = 88.86%
5. SVM = 90.30%  

In [249]:
# random forest and logistic regression is give best accuracy